internal/controllers/flowcontrol/synthesis.go

```diff
@@ -0,0 +1,81 @@
+package flowcontrol
+
+import (
+	"context"
+	"fmt"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type synthesisConcurrencyLimiter struct {
+	client client.Client
+	limit  int
+}
+
+func NewSynthesisConcurrencyLimiter(mgr ctrl.Manager, limit int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("synthesisConcurrencyLimiter").
+		Watches(&apiv1.Composition{}, manager.SingleEventHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "synthesisConcurrencyLimiter")).
+		Complete(&synthesisConcurrencyLimiter{
+			client: mgr.GetClient(),
+			limit:  limit,
+		})
+}
+
+func (c *synthesisConcurrencyLimiter) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	list := &apiv1.CompositionList{}
+	err := c.client.List(ctx, list)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var active int
+	var pending int
+	var next *apiv1.Composition
+	for _, comp := range list.Items {
+		comp := comp
```

Comment:
Are we on go 1.22? If so this is no longer needed

Reply:
Still on 1.21 I believe, good reminder to update :)